api/src/main/java/org/openmrs/obs/handler/AbstractHandler.java

```diff
@@ -34,124 +40,148 @@ public class AbstractHandler {
-	 *             instead.
+	 * @see org.openmrs.obs.ComplexObsHandler#getObs(Obs, String)
 	 */
-	@Deprecated
-	public String getExtension(String filename) {
-		String result = FilenameUtils.getExtension(filename);
+	public Obs getObs(Obs obs, String view) {
+		String key = parseNewOrLegacyKey(obs);
 		
-		if (StringUtils.isEmpty(result)) {
-			result = FilenameUtils.getBaseName(filename);
-			
-			if (StringUtils.isEmpty(result)) {
-				result = "raw";
-			}
+		byte[] bytes;
+		try (InputStream is = storageService.getData(key)) {
+			bytes = IOUtils.toByteArray(is);
+		}
+		catch (IOException e) {
+			throw new UncheckedIOException(e);
 		}
 		
-		return result;
+		ComplexData complexData = new ComplexData(parseDataTitle(obs), bytes);
+		injectMissingMetadata(key, complexData);
+		obs.setComplexData(complexData);
+		return obs;
 	}
 	
-	/**
-	 * @see org.openmrs.obs.ComplexObsHandler#getObs(Obs, String)
-	 */
-	public Obs getObs(Obs obs, String view) {
-		File file = BinaryDataHandler.getComplexDataFile(obs);
-		log.debug("value complex: " + obs.getValueComplex());
-		log.debug("file path: " + file.getAbsolutePath());
-		ComplexData complexData = null;
+	public Obs saveObs(Obs obs) throws APIException {
 		try {
-			complexData = new ComplexData(file.getName(), OpenmrsUtil.getFileAsBytes(file));
+			byte[] data = (byte[]) obs.getComplexData().getData();
+			
+			String key = storageService.saveData(outputStream -> {
+				IOUtils.write( data, outputStream);
+			}, ObjectMetadata.builder().setLength((long) data.length).build(), getObsDir());
+			// Store the filename in the Obs
+			obs.setValueComplex(StringUtils.defaultIfBlank(obs.getComplexData().getTitle(), key) + "|" + key);
+			obs.setComplexData(null);
 		}
-		catch (IOException e) {
-			log.error("Trying to read file: " + file.getAbsolutePath(), e);
+		catch (Exception e) {
+			throw new APIException("Obs.error.writing.binary.data.complex", null, e);
 		}
-		String mimeType = OpenmrsUtil.getFileMimeType(file);
-		complexData.setMimeType(mimeType);
-		obs.setComplexData(complexData);
-		
+
 		return obs;
 	}
 	
 	/**
 	 * @see org.openmrs.obs.ComplexObsHandler#purgeComplexData(org.openmrs.Obs)
 	 */
 	public boolean purgeComplexData(Obs obs) {
-		File file = getComplexDataFile(obs);
-		if (!file.exists()) {
-			return true;
-		} else if (file.delete()) {
+		String key = parseNewOrLegacyKey(obs);
+		
+		try {
+			storageService.purgeData(key);
 			obs.setComplexData(null);
 			return true;
+		} catch (IOException e) {
+			log.warn("Could not delete complex data object for obsId={} located at {}", obs.getObsId(), key);
+			return false;
 		}
-		
-		log.warn(
-		    "Could not delete complex data object for obsId=" + obs.getObsId() + " located at " + file.getAbsolutePath());
-		return false;
 	}
-	
+
+	/**
+	 * Should be replaced with just parseDataKey once migration is over.
+	 * 
+	 * @param obs complex obs
+	 * @return key
+	 * @deprecated 2.8.0
+	 * @since 2.8.0
+	 */
+	@Deprecated
```

Comment:
https://openmrs.atlassian.net/wiki/spaces/docs/pages/25477044/Java+Conventions#Deprecation

Reply:
I'll just add the logic to parseDataKey and remove this method altogether.